File: iOS/Spaces/Models/UserProfile.swift
import Foundation

struct UserProfile: Identifiable, Hashable {
    let id: String
    let uid: String
    let displayName: String
    let emojiAvatar: String
    let statusMessage: String
    let email: String?
    let phoneNumber: String?
    let linkedProviders: [LinkedProvider]
    let blockedUsers: [BlockedUser]

    init(
        id: String,
        uid: String,
        displayName: String,
        emojiAvatar: String,
        statusMessage: String,
        email: String?,
        phoneNumber: String?,
        linkedProviders: [LinkedProvider],
        blockedUsers: [BlockedUser] = []
    ) {
        self.id = id
        self.uid = uid
        self.displayName = displayName
        self.emojiAvatar = emojiAvatar
        self.statusMessage = statusMessage
        self.email = email
        self.phoneNumber = phoneNumber
        self.linkedProviders = linkedProviders
        self.blockedUsers = blockedUsers
    }
}

struct AccountDevice: Identifiable, Hashable {
    let id: String
    let deviceID: String
    let platform: String
    let publicKey: String?
    let createdAt: Date?
    let lastActiveAt: Date?
}

struct PushTokenRecord: Identifiable, Hashable {
    let id: String
    let userID: String
    let token: String
    let platform: String
    let deviceID: String?
    let enabled: Bool
    let createdAt: Date?
    let updatedAt: Date?
}

struct BlockedUser: Identifiable, Hashable {
    let id: String
    let uid: String
    let displayName: String
    let emojiAvatar: String
    let blockedAt: Date?
}

enum LinkedProvider: String, CaseIterable, Identifiable, Hashable {
    case apple = "Apple"
    case google = "Google"
    case phone = "Phone"

    var id: String { rawValue }

    init?(firebaseProviderID: String) {
        switch firebaseProviderID {
        case "apple.com":
            self = .apple
        case "google.com":
            self = .google
        case "phone":
            self = .phone
        default:
            return nil
        }
    }
}

File: Android/app/src/main/java/com/arcinteractive/spaces/data/model/UserProfile.kt
package com.arcinteractive.spaces.data.model

data class UserProfile(
    val id: String,
    val uid: String,
    val displayName: String,
    val emojiAvatar: String,
    val statusMessage: String,
    val email: String?,
    val phoneNumber: String?,
    val linkedProviders: List<LinkedProvider>,
    val blockedUsers: List<BlockedUser> = emptyList()
)

data class BlockedUser(
    val id: String,
    val uid: String,
    val displayName: String,
    val emojiAvatar: String,
    val blockedAt: java.util.Date?
)

data class AccountDevice(
    val id: String,
    val deviceId: String,
    val platform: String,
    val publicKey: String?,
    val createdAt: java.util.Date?,
    val lastActiveAt: java.util.Date?
)

data class PushTokenRecord(
    val id: String,
    val userId: String,
    val token: String,
    val platform: String,
    val deviceId: String?,
    val enabled: Boolean,
    val createdAt: java.util.Date?,
    val updatedAt: java.util.Date?
)

enum class LinkedProvider(
    val label: String
) {
    Apple("Apple"),
    Google("Google"),
    Phone("Phone");

    companion object {
        fun fromFirebaseProviderId(providerId: String): LinkedProvider? = when (providerId) {
            "google.com" -> Google
            "phone" -> Phone
            "apple.com" -> Apple
            else -> null
        }
    }
}
